## Netflix Data Analysis

### Overview:

The aim of this case study is to load and prepare the data using Pandas and visualise the data in PowerBI to extract key insights

#### Importing necessary Libraries

In [1]:
# Importing the required libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### Changing display settings

In [2]:
# Modifying settings to ensure that all the required rows and columns are displayed
pd.set_option('display.max_columns', 5000)
pd.set_option('display.max_rows', 5000)

### Reading the data

### <font color = blue> Option 1: Downloading the data directly from Kaggle

In [3]:
import os
from kaggle.api.kaggle_api_extended import KaggleApi

# Initialize the Kaggle API
api = KaggleApi()
api.authenticate()

# Define the dataset and download path
dataset = 'shivamb/netflix-shows'
download_path = 'netflix_dataset'

# Create the download directory if it doesn't exist
os.makedirs(download_path, exist_ok=True)

# Download the dataset
api.dataset_download_files(dataset, path=download_path, unzip=True)

Dataset URL: https://www.kaggle.com/datasets/shivamb/netflix-shows


In [4]:
import pandas as pd
import os

# Define the path to the CSV file
csv_file = os.path.join(download_path, 'netflix_titles.csv')

# Load the CSV into a DataFrame
df = pd.read_csv(csv_file)

# Display the first few rows
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

### <font color = blue> Option 2: Reading the data from the local folder

In [5]:
df = pd.read_csv('netflix_titles.csv')

In [6]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

## Understanding and fixing the data

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### Null checks

In [8]:
# Calculating the null percentage for everycolumn in dataframe - 

# Creating a dataframe of null percentages for current_app
nullperc_df  = pd.DataFrame((df.isnull().sum())*100/df.shape[0])

# Formatting the dataframe of null percentages 
nullperc_df.reset_index(inplace = True)
nullperc_df  = nullperc_df.rename(columns = {'index':'Column Name',0:'Null Percentage'})
nullperc_df=nullperc_df[nullperc_df['Null Percentage']>0]

# Displaying the dataframe of null percentages for current_app:
nullperc_df

Column Name  Null Percentage
3    director        29.908028
4        cast         9.367549
5     country         9.435676
6  date_added         0.113546
8      rating         0.045418
9    duration         0.034064

In [9]:
df.fillna({'director':'Unknown', 'cast':'Unknown','country':'Unknown'},inplace = True)

In [10]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [11]:
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water          Unknown   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...

### Removing double quotes and newline characters from the description column

In [12]:
df['description'] = df['description'].str.replace('\n', '', regex=False)
df['description'] = df['description'].str.replace('"', '', regex=False)

### Removing double quotes and newline characters from the title column

In [13]:
df['title'] = df['title'].str.replace('\n', '', regex=False)
df['title'] = df['title'].str.replace('"', '', regex=False)

### Convert the datatype of date_added

In [14]:
df['date_added'] = pd.to_datetime(df['date_added'])
df.head(3)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water          Unknown   
2      s3  TV Show             Ganglands  Julien Leclercq   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...        Unknown   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...

### Adding two new columns - Year added and month added

In [15]:
df['year_added'] = pd.DatetimeIndex(df['date_added']).year
df['month_added'] = pd.DatetimeIndex(df['date_added']).month_name()
df.head(3)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water          Unknown   
2      s3  TV Show             Ganglands  Julien Leclercq   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...        Unknown   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  year_added month_added  
0  As her father nears the end of his life, filmm...      2021.0   September  
1  After crossing paths at a party, a Cape Town t...      2021.0   September  
2  To protect his family from a powerful drug lor...      2021.0   September

### Fixing nulls in date_added

In [16]:
df.loc[df.date_added.isna()].head()

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories  Unknown   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown  Unknown   
6795   s6796  TV Show                                      Frasier  Unknown   
6806   s6807  TV Show                                      Friends  Unknown   
6901   s6902  TV Show                              Gunslinger Girl  Unknown   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   

     date_added  release_year rating    duration  \
6066        NaT          2013  TV-MA   2 Seasons   
6174        NaT          2018  TV-PG   5 Seasons   
6795        NaT          2003  TV-PG  11 Seasons   
6806        NaT          2003  TV-14  10 Seasons   
6901        NaT          2008  TV-14   2 Seasons   

                                     listed_in  \
6066  British TV Shows, TV Comedies, TV Dramas   
6174                                Docuseries   
6795            Classic & Cult TV, TV Comedies   
6806            Classic & Cult TV, TV Comedies   
6901              Anime Series, Crime TV Shows   

                                            description  year_added  \
6066  Set during the Russian Revolution, this comic ...         NaN   
6174  This CNN original series has chef Anthony Bour...         NaN   
6795  Frasier Crane is a snooty but lovable Seattle ...         NaN   
6806  This hit sitcom follows the merry misadventure...         NaN   
6901  On the surface, the Social Welfare Agency appe...         NaN   

     month_added  
6066         NaN  
6174         NaN  
6795         NaN  
6806         NaN  
6901         NaN

### Replacing the null year_added with a the average time it takes to add a show

In [17]:
np.mean(df['year_added'] - df['release_year'] )

4.688416505626918

##### On an average it takes 4.6884165 years after release to get the show added to Netflix - Rounding this down to 4

In [18]:
# Assigning date_added as release year+4 wherever it is null
df['date_added'] = df.apply(
    (lambda x: pd.to_datetime(x['release_year']+4, format = '%Y') if pd.isna(x['date_added'])  else x['date_added']), axis = 1)

In [19]:
df['year_added'] = pd.DatetimeIndex(df['date_added']).year
df['month_added'] = pd.DatetimeIndex(df['date_added']).month_name()

### Unexpected data quality issue: Rating column has durations

In [20]:
df.rating.unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', '74 min', '84 min', '66 min', 'NR', nan,
       'TV-Y7-FV', 'UR'], dtype=object)

In [21]:
def is_duration(val):
    return 'min' in val.split()

df['duration'] = df.apply(
    (lambda x: x['rating'] if (pd.isna(x['duration']) and is_duration(x['rating'])) else x['duration']), axis = 1)

In [22]:
df.loc[df.rating.isin(['84 min', '74 min', '66 min'])].duration

5541    74 min
5794    84 min
5813    66 min
Name: duration, dtype: object

In [23]:
# Duration column is fixed, fixing the rating column:
df['rating'] = df.apply(
    (lambda x: 'Unknown' if (x['rating'] == x['duration']) else x['rating']), axis = 1)

In [24]:
df['rating'].unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', 'Unknown', 'NR', nan, 'TV-Y7-FV', 'UR'],
      dtype=object)

In [25]:
df.fillna({'rating':'Unknown'},inplace = True)

### Adding the cast into a deparate dataframe

In [46]:
cast_df = df[['show_id','cast']]
cast_df.set_index('show_id', inplace = True)
cast_df = cast_df.cast.apply(lambda x: x.split(',')).explode().reset_index()

#Stripping white spaces and commas
cast_df.cast = cast_df.cast.str.strip(', ')

In [47]:
cast_df

show_id                   cast
0          s1                Unknown
1          s2             Ama Qamata
2          s2            Khosi Ngema
3          s2          Gail Mabalane
4          s2         Thabang Molaba
...       ...                    ...
64946   s8807       Manish Chaudhary
64947   s8807           Meghna Malik
64948   s8807          Malkeet Rauni
64949   s8807         Anita Shabdish
64950   s8807  Chittaranjan Tripathy

[64951 rows x 2 columns]

In [48]:
cast_df.to_csv('netflix_cast.csv')

### Splitting Genre into different df

In [40]:
genre_df = df[['show_id','listed_in']]
genre_df.set_index('show_id', inplace = True)
genre_df = genre_df.listed_in.apply(lambda x: x.split(',')).explode().reset_index()

#Stripping white spaces and commas
genre_df.listed_in = genre_df.listed_in.str.strip(', ')

In [41]:
genre_df

show_id                 listed_in
0          s1             Documentaries
1          s2    International TV Shows
2          s2                 TV Dramas
3          s2              TV Mysteries
4          s3            Crime TV Shows
...       ...                       ...
19318   s8806  Children & Family Movies
19319   s8806                  Comedies
19320   s8807                    Dramas
19321   s8807      International Movies
19322   s8807          Music & Musicals

[19323 rows x 2 columns]

In [42]:
genre_df.to_csv('netflix_genres.csv')

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3   director      8807 non-null   object        
 4   cast          8807 non-null   object        
 5   country       8807 non-null   object        
 6   date_added    8807 non-null   datetime64[ns]
 7   release_year  8807 non-null   int64         
 8   rating        8807 non-null   object        
 9   duration      8807 non-null   object        
 10  listed_in     8807 non-null   object        
 11  description   8807 non-null   object        
 12  year_added    8807 non-null   int64         
 13  month_added   8807 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(11)
memory usage: 963.4+ KB


### Splitting Country into different df

In [43]:
country_df = df[['show_id','country']]
country_df.set_index('show_id', inplace = True)
country_df = country_df.country.apply(lambda x: x.split(',')).explode().reset_index()

#Stripping white spaces and commas
country_df.country = country_df.country.str.strip(', ')

In [44]:
country_df

show_id        country
0          s1  United States
1          s2   South Africa
2          s3        Unknown
3          s4        Unknown
4          s5          India
...       ...            ...
10845   s8803  United States
10846   s8804        Unknown
10847   s8805  United States
10848   s8806  United States
10849   s8807          India

[10850 rows x 2 columns]

In [45]:
country_df.to_csv('netflix_countries.csv')

### Splitting movies and series into different dataframes

In [36]:
df_movies = df.loc[df.type =='Movie']
df_movies['duration_min'] = df_movies['duration'].apply(lambda x: float(x.split()[0]))
df_movies

<ipython-input-36-e0e695294a0e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_movies['duration_min'] = df_movies['duration'].apply(lambda x: float(x.split()[0]))


show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                               Unknown   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                               Unknown 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...        2021   
6    

In [37]:
df_movies.to_csv('netflix_movies.csv')

In [38]:
df_shows = df.loc[df.type =='TV Show']
df_shows.to_csv('netflix_shows.csv')

## Writing the cleaned data to a csv

In [39]:
df.to_csv('netflix_cleaned.csv')